compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+  mli.emitServerPrelude();
+
+  forv_Vec(ModuleSymbol, md, mds) {
+    mli.emit(md);
+  }
+
+  mli.emitMarshalRoutines();
+  mli.emitServerDispatchRoutine();
+
+  return;
+}
+
+MLIContext::MLIContext(bool injectDebugPrintlines, bool separateHeaders) {
+
+  // Yes, I know this isn't the most optimal way to initialize these!
+  this->injectDebugPrintlines = injectDebugPrintlines;
+  this->separateHeaders = separateHeaders;
+
+  openCFile(&this->fiMarshalling, gen_mli_marshalling, "c");
+  openCFile(&this->fiClientBundle, gen_mli_client_bundle, "c");
+  openCFile(&this->fiServerBundle, gen_mli_server_bundle, "c");
+  this->info = gGenInfo;
+
+  return;
+}
+
+MLIContext::~MLIContext() {
+
+  // Now, turn beautify ON!
+  closeCFile(&this->fiMarshalling, true);
+  closeCFile(&this->fiClientBundle, true);
+  closeCFile(&this->fiServerBundle, true);
+
+  return;
+}
+
+bool MLIContext::shouldEmit(ModuleSymbol* md) {
+  return (md->modTag != MOD_INTERNAL);
+}
+
+bool MLIContext::shouldEmit(FnSymbol* fn) {
+  return (fn->hasFlag(FLAG_EXPORT) && not fn->hasFlag(FLAG_GEN_MAIN_FUNC));
+}
+
+void MLIContext::emit(ModuleSymbol* md) {
+  if (not this->shouldEmit(md)) { return; }
+
+  const std::vector<FnSymbol*> fns = md->getTopLevelFunctions(true);
+
+  for_vector(FnSymbol, fn, fns) {
+    if (not this->shouldEmit(fn)) { continue; }
+
+    this->exps.push_back(fn);
+    this->emit(fn);
+  }
+
+  return;
+}
+
+void MLIContext::emit(FnSymbol* fn) {
+  if (not this->shouldEmit(fn)) { return; }
+
+  this->verifyPrototype(fn); 
+  this->emitClientWrapper(fn);
+  this->emitServerWrapper(fn);
+
+  return;
+}
+
+void MLIContext::emitClientPrelude(void) {
+  std::string gen;
+
+  // TODO: Trying to work around the Travis malloc complaint.
+  gen += "#define CHPL_MLI_IS_CLIENT_\n";
+  gen += this->genHeaderInc("chpl_mli_marshalling.c");
+  gen += this->genNote("We use Makefile magic to make this visible!");
+  gen += this->genHeaderInc("mli_client_runtime.c");
+  gen += "\n";
+
+  this->setOutputAndWrite(&this->fiClientBundle, gen);
+
+  return;
+}
+
+std::string MLIContext::genHeaderInc(const char* header, bool system) {
+  std::string gen;
+
+  gen += "#include ";
+  gen += system ? "<" : "\"";
+  gen += header;
+  gen += system ? ">" : "\"";
+  gen += "\n";
+
+  return gen;
+}
+
+std::string MLIContext::genComment(const char* msg, const char* pfx) {
+  std::string gen;
+
+  gen += "// ";
+  gen += pfx;
+  gen += pfx != "" ? ": " : "";
+  gen += msg;
+  gen += "\n";
+
+  return gen;
+}
+
+std::string MLIContext::genNote(const char* msg) {
+  return this->genComment(msg, "NOTE");
+}
+
+std::string MLIContext::genTodo(const char* msg) {
+  return this->genComment(msg, "TODO");
+}
+
+std::string MLIContext::genFuncToSetServerGlobals(void) {
+  std::string gen;
+
+  gen += "void chpl_mli_server_set_conf(void)";
+  gen += scope_begin;
+  gen += "chpl_server_conf.debug=";
+  gen += this->injectDebugPrintlines ? "1" : "0";
+  gen += ";\n";
+  gen += scope_end;
```

Comment:
It seems like you kinda already do that, so I'm not sure what the point of inserting this code is

Reply:
The server listen loop isn't generated, so it needs to have a runtime flag to determine whether or not to print some items. I just made it so the value of that flag is turned on or off when the codegen flag is set.

Reply:
Ah, I see.  I think if possible, it would be better to use a `#define` for the debugging level and surround the debug print calls in the server listen loop with an `#ifdef` on the `#define`d variable.  That way an optimized version of the code won't even have to pay the cost of stepping into the inserted if statement on server runtime line 82

Reply:
Good idea.

Reply:
I'll work on that after I satisfy Travis and clean up formatting issues.